File: app/src/main/java/com/example/mutidemo/ui/RefreshAndLoadMoreActivity.java
package com.example.mutidemo.ui;

import android.content.Intent;
import android.os.Handler;

import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.mutidemo.adapter.NewsAdapter;
import com.example.mutidemo.bean.NewsListBean;
import com.example.mutidemo.databinding.ActivityRefreshBinding;
import com.example.mutidemo.mvp.presenter.NewsListPresenterImpl;
import com.example.mutidemo.mvp.view.INewsListView;
import com.example.mutidemo.util.OtherUtils;
import com.pengxh.androidx.lite.base.AndroidxBaseActivity;
import com.pengxh.androidx.lite.utils.WeakReferenceHandler;

import java.util.ArrayList;
import java.util.List;

/**
 * @author: Pengxh
 * @email: 290677893@qq.com
 * @description: TODO
 * @date: 2020/2/21 19:16
 */
public class RefreshAndLoadMoreActivity extends AndroidxBaseActivity<ActivityRefreshBinding> implements INewsListView {

    /**
     * 设置一个集合，用来存储网络请求到的数据
     */
    private List<NewsListBean.ResultBeanX.ResultBean.ListBean> dataBeans = new ArrayList<>();
    /**
     * 自定义刷新和加载的标识，默认为false
     */
    private boolean isRefresh, isLoadMore = false;
    /**
     * 起始页
     */
    private int defaultPage = 0;
    private NewsAdapter newsAdapter;
    private NewsListPresenterImpl newsPresenter;
    private static WeakReferenceHandler weakReferenceHandler;

    @Override
    protected void setupTopBarLayout() {

    }

    @Override
    public void initData() {
        weakReferenceHandler = new WeakReferenceHandler(callback);
        newsPresenter = new NewsListPresenterImpl(this);
        newsPresenter.onReadyRetrofitRequest("娱乐", defaultPage);
    }

    @Override
    public void initEvent() {
        viewBinding.refreshLayout.setOnRefreshListener(refreshLayout -> {
            isRefresh = true;
            //刷新之后页码重置
            defaultPage = 0;
            newsPresenter.onReadyRetrofitRequest("娱乐", defaultPage);
        });
        viewBinding.refreshLayout.setOnLoadMoreListener(refreshLayout -> {
            isLoadMore = true;
            defaultPage++;
            newsPresenter.onReadyRetrofitRequest("娱乐", defaultPage);
        });
    }

    private final Handler.Callback callback = msg -> {
        if (msg.what == 10000) {
            if (isRefresh || isLoadMore) {
                newsAdapter.notifyDataSetChanged();
            } else {
                //首次加载数据
                newsAdapter = new NewsAdapter(this, dataBeans);
                viewBinding.newsRecyclerView.setLayoutManager(new LinearLayoutManager(this));
                viewBinding.newsRecyclerView.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));
                viewBinding.newsRecyclerView.setAdapter(newsAdapter);
                newsAdapter.setOnNewsItemClickListener(position -> {
                    NewsListBean.ResultBeanX.ResultBean.ListBean bean = dataBeans.get(position);
                    Intent intent = new Intent(this, NewsDetailsActivity.class);
                    intent.putExtra("title", bean.getTitle());
                    intent.putExtra("src", bean.getSrc());
                    intent.putExtra("time", bean.getTime());
                    intent.putExtra("content", bean.getContent());
                    startActivity(intent);
                });
            }
        }
        return true;
    };

    @Override
    public void showProgress() {
        if (!isRefresh && !isLoadMore) {
            OtherUtils.showLoadingDialog(this, "加载数据中，请稍后...");
        }
    }

    @Override
    public void hideProgress() {
        OtherUtils.dismissLoadingDialog();
    }

    @Override
    public void showNetWorkData(NewsListBean response) {
        if (response.getCode().equals("10000")) {
            List<NewsListBean.ResultBeanX.ResultBean.ListBean> listBeans = response.getResult().getResult().getList();
            if (isRefresh) {
                dataBeans.clear();//下拉刷新必须先清空之前的List，不然会出现数据重复的问题
                dataBeans = listBeans;
                viewBinding.refreshLayout.finishRefresh();
                isRefresh = false;
            } else if (isLoadMore) {
                dataBeans.addAll(listBeans);
                viewBinding.refreshLayout.finishLoadMore();
                isLoadMore = false;
            } else {
                dataBeans = listBeans;
            }
            weakReferenceHandler.sendEmptyMessage(10000);
        }
    }

    @Override
    public void onDestroy() {
        if (newsPresenter != null) {
            newsPresenter.disposeRetrofitRequest();
        }
        super.onDestroy();
    }
}

File: app/src/main/java/com/example/mutidemo/view/RefreshAndLoadMoreActivity.kt
package com.example.mutidemo.view

import android.os.Handler
import android.os.Message
import android.view.View
import androidx.lifecycle.ViewModelProvider
import com.example.mutidemo.R
import com.example.mutidemo.extensions.addAll
import com.example.mutidemo.model.NewsListModel
import com.example.mutidemo.util.LoadingDialogHub
import com.example.mutidemo.vm.NewsViewModel
import com.pengxh.kt.lite.adapter.NormalRecyclerAdapter
import com.pengxh.kt.lite.adapter.ViewHolder
import com.pengxh.kt.lite.base.KotlinBaseActivity
import com.pengxh.kt.lite.divider.ItemDecoration
import com.pengxh.kt.lite.extensions.dp2px
import com.pengxh.kt.lite.extensions.navigatePageTo
import com.pengxh.kt.lite.extensions.show
import com.pengxh.kt.lite.utils.WeakReferenceHandler
import com.pengxh.kt.lite.vm.LoadState
import kotlinx.android.synthetic.main.activity_refresh.*
import kotlinx.android.synthetic.main.item_news_rv_l.*

/**
 * @author: Pengxh
 * @email: 290677893@qq.com
 * @date: 2020/2/21 19:16
 */
class RefreshAndLoadMoreActivity : KotlinBaseActivity() {

    private lateinit var weakReferenceHandler: WeakReferenceHandler
    private lateinit var newsViewModel: NewsViewModel
    private lateinit var newsAdapter: NormalRecyclerAdapter<NewsListModel.ResultBeanX.ResultBean.ListBean>
    private var dataBeans: MutableList<NewsListModel.ResultBeanX.ResultBean.ListBean> = ArrayList()
    private var isRefresh = false
    private var isLoadMore = false
    private var pageIndex = 0

    override fun setupTopBarLayout() {}

    override fun initLayoutView(): Int = R.layout.activity_refresh

    override fun initData() {
        weakReferenceHandler = WeakReferenceHandler(callback)
        newsViewModel = ViewModelProvider(this)[NewsViewModel::class.java]
        newsViewModel.resultModel.observe(this) {
            if (it.code == "10000") {
                val dataRows = it.result.result.list
                when {
                    isRefresh -> {
                        dataBeans.clear()
                        dataBeans = dataRows
                        refreshLayout.finishRefresh()
                        isRefresh = false
                    }
                    isLoadMore -> {
                        if (dataRows.size == 0) {
                            "到底了，别拉了".show(this)
                        }
                        dataBeans.addAll(dataRows)
                        refreshLayout.finishLoadMore()
                        isLoadMore = false
                    }
                    //首次加载数据
                    else -> dataBeans = dataRows
                }
                weakReferenceHandler.sendEmptyMessage(2023031301)
            }
        }
    }

    override fun initEvent() {
        refreshLayout.setOnRefreshListener {
            isRefresh = true
            //刷新之后页码重置
            pageIndex = 0
            getNewsList()
        }
        refreshLayout.setOnLoadMoreListener {
            isLoadMore = true
            pageIndex++
            getNewsList()
        }
    }

    override fun observeRequestState() {
        newsViewModel.loadState.observe(this) {
            when (it) {
                LoadState.Loading -> {
                    if (!isRefresh && !isLoadMore) {
                        LoadingDialogHub.show(this, "加载数据中，请稍后...")
                    }
                }
                else -> LoadingDialogHub.dismiss()
            }
        }
    }

    override fun onResume() {
        super.onResume()
        pageIndex = 1
        getNewsList()
    }

    private fun getNewsList() {
        newsViewModel.getNewsList("头条", pageIndex)
    }

    private val callback = Handler.Callback { msg: Message ->
        if (msg.what == 2023031301) {
            newsAdapter =
                object : NormalRecyclerAdapter<NewsListModel.ResultBeanX.ResultBean.ListBean>(
                    R.layout.item_news_rv_l, dataBeans
                ) {
                    override fun convertView(
                        viewHolder: ViewHolder, position: Int,
                        item: NewsListModel.ResultBeanX.ResultBean.ListBean
                    ) {
                        val img: String = item.pic
                        if (img == "" || img.endsWith(".gif")) {
                            newsPicture.visibility = View.GONE
                        } else {
                            viewHolder.setImageResource(R.id.newsPicture, img)
                        }

                        viewHolder.setText(R.id.newsTitle, item.title)
                            .setText(R.id.newsSrc, item.src)
                            .setText(R.id.newsTime, item.time)
                    }
                }
            newsRecyclerView.addItemDecoration(
                ItemDecoration(0f, 130f.dp2px(this@RefreshAndLoadMoreActivity).toFloat())
            )
            newsRecyclerView.adapter = newsAdapter
            newsAdapter.setOnItemClickedListener(object :
                NormalRecyclerAdapter.OnItemClickedListener<NewsListModel.ResultBeanX.ResultBean.ListBean> {
                override fun onItemClicked(
                    position: Int, t: NewsListModel.ResultBeanX.ResultBean.ListBean
                ) {
                    navigatePageTo<NewsDetailsActivity>(addAll(t.title, t.src, t.time, t.content))
                }
            })
        }
        true
    }
}